File: imagepipeline/src/test/java/com/facebook/imagepipeline/memory/PooledByteArrayBufferedInputStreamTest.java
/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

package com.facebook.imagepipeline.memory;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;

import com.facebook.common.references.ResourceReleaser;
import org.robolectric.RobolectricTestRunner;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.mockito.Mockito.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public class PooledByteArrayBufferedInputStreamTest {

  private ResourceReleaser mResourceReleaser;
  private byte[] mBuffer;
  private PooledByteArrayBufferedInputStream mPooledByteArrayBufferedInputStream;

  @Before
  public void setUp() {
    mResourceReleaser = mock(ResourceReleaser.class);
    final byte[] bytes = new byte[256];
    for (int i = 0; i < 256; ++i) {
      bytes[i] = (byte) i;
    }
    InputStream unbufferedStream = new ByteArrayInputStream(bytes);
    mBuffer = new byte[10];
    mPooledByteArrayBufferedInputStream = new PooledByteArrayBufferedInputStream(
        unbufferedStream,
        mBuffer,
        mResourceReleaser);
  }

  @Test
  public void testSingleByteRead() throws IOException {
    for (int i = 0; i < 256; ++i) {
      assertEquals(i, mPooledByteArrayBufferedInputStream.read());
    }
    assertEquals(-1, mPooledByteArrayBufferedInputStream.read());
  }

  @Test
  public void testReleaseOnClose() throws IOException {
    mPooledByteArrayBufferedInputStream.close();
    verify(mResourceReleaser).release(mBuffer);
    mPooledByteArrayBufferedInputStream.close();
    // we do not expect second close to release resource again,
    // the one checked bellow is the one that happened when close was called for the first time
    verify(mResourceReleaser).release(any(byte[].class));
  }

  @Test
  public void testSkip() throws IOException {
    // buffer some data
    mPooledByteArrayBufferedInputStream.read();
    assertEquals(99, mPooledByteArrayBufferedInputStream.skip(99));
    assertEquals(100, mPooledByteArrayBufferedInputStream.read());
  }

  @Test
  public void testSkip2() throws IOException {
    int i = 0;
    while (i < 256) {
      assertEquals(i, mPooledByteArrayBufferedInputStream.read());
      i += mPooledByteArrayBufferedInputStream.skip(7) + 1;
    }
  }

  @Test
  public void testMark() {
    assertFalse(mPooledByteArrayBufferedInputStream.markSupported());
  }

  @Test
  public void testReadWithByteArray() throws IOException {
    byte[] readBuffer = new byte[5];
    assertEquals(5, mPooledByteArrayBufferedInputStream.read(readBuffer));
    assertFilledWithConsecutiveBytes(readBuffer, 0, 5, 0);
  }

  @Test
  public void testNonFullRead() throws IOException {
    byte[] readBuffer = new byte[200];
    assertEquals(10, mPooledByteArrayBufferedInputStream.read(readBuffer));
    assertFilledWithConsecutiveBytes(readBuffer, 0, 10, 0);
    assertFilledWithZeros(readBuffer, 10, 200);
  }

  @Test
  public void testNonFullReadWithOffset() throws IOException {
    byte[] readBuffer = new byte[200];
    assertEquals(10, mPooledByteArrayBufferedInputStream.read(readBuffer, 45, 75));
    assertFilledWithZeros(readBuffer, 0, 45);
    assertFilledWithConsecutiveBytes(readBuffer, 45, 55, 0);
    assertFilledWithZeros(readBuffer, 55, 200);
  }

  @Test
 public void testReadsCombined() throws IOException {
    byte[] readBuffer = new byte[5];
    int i = 0;
    while (i <= 245) {
      assertEquals(i, mPooledByteArrayBufferedInputStream.read());

      assertEquals(5, mPooledByteArrayBufferedInputStream.read(readBuffer));
      assertFilledWithConsecutiveBytes(readBuffer, 0, readBuffer.length, i + 1);

      assertEquals(3, mPooledByteArrayBufferedInputStream.read(readBuffer, 1, 3));
      assertEquals((byte) (i + 1), readBuffer[0]);
      assertFilledWithConsecutiveBytes(readBuffer, 1, 4, i + 6);
      assertEquals((byte) (i + 5), readBuffer[4]);

      assertEquals(2, mPooledByteArrayBufferedInputStream.skip(2));

      i += 11;
    }

    assertEquals(256 - i, mPooledByteArrayBufferedInputStream.available());
  }

  /**
   * Given byte array, asserts that bytes in [startOffset, endOffset) range are all zeroed;
   *
   * @param byteArray
   * @param startOffset
   * @param endOffset
   */
  private static void assertFilledWithZeros(
      final byte[] byteArray,
      final int startOffset,
      final int endOffset) {
    for (int i = startOffset; i < endOffset; ++i) {
      assertEquals(0, byteArray[i]);
    }
  }

  /**
   * Given byte array, asserts that each byte in (startOffset, endOffset) range has value equal
   * to value of previous byte plus one (mod 255) and byteArray[startOffset] is equal to firstByte.
   *
   * @param byteArray
   * @param startOffset
   * @param endOffset
   * @param firstByte
   */
  private static void assertFilledWithConsecutiveBytes(
      final byte[] byteArray,
      final int startOffset,
      final int endOffset,
      int firstByte) {
    for (int i = startOffset; i < endOffset; ++i) {
      assertEquals((byte) firstByte++, byteArray[i]);
    }
  }
}

File: imagepipeline/src/test/java/com/facebook/common/memory/PooledByteArrayBufferedInputStreamTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.common.memory

import com.facebook.common.references.ResourceReleaser
import java.io.ByteArrayInputStream
import java.io.IOException
import org.assertj.core.api.Assertions
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class PooledByteArrayBufferedInputStreamTest {

  private lateinit var resourceReleaser: ResourceReleaser<ByteArray>
  private lateinit var buffer: ByteArray
  private lateinit var pooledByteArrayBufferedInputStream: PooledByteArrayBufferedInputStream

  @Before
  fun setUp() {
    @Suppress("UNCHECKED_CAST")
    resourceReleaser = Mockito.mock(ResourceReleaser::class.java) as ResourceReleaser<ByteArray>
    val bytes = ByteArray(256)
    for (i in 0 until 256) {
      bytes[i] = i.toByte()
    }
    val unbufferedStream = ByteArrayInputStream(bytes)
    buffer = ByteArray(10)
    pooledByteArrayBufferedInputStream =
        PooledByteArrayBufferedInputStream(unbufferedStream, buffer, resourceReleaser)
  }

  @Test
  @Throws(IOException::class)
  fun testSingleByteRead() {
    for (i in 0 until 256) {
      Assertions.assertThat(pooledByteArrayBufferedInputStream.read()).isEqualTo(i)
    }
    Assertions.assertThat(pooledByteArrayBufferedInputStream.read()).isEqualTo(-1)
  }

  @Test
  @Throws(IOException::class)
  fun testReleaseOnClose() {
    pooledByteArrayBufferedInputStream.close()
    Mockito.verify(resourceReleaser).release(buffer)
    pooledByteArrayBufferedInputStream.close()
    // we do not expect second close to release resource again,
    // the one checked bellow is the one that happened when close was called for the first time
    Mockito.verify(resourceReleaser, Mockito.times(1)).release(Mockito.any(ByteArray::class.java))
  }

  @Test
  @Throws(IOException::class)
  fun testSkip() {
    // buffer some data
    pooledByteArrayBufferedInputStream.read()
    Assertions.assertThat(pooledByteArrayBufferedInputStream.skip(99)).isEqualTo(99)
    Assertions.assertThat(pooledByteArrayBufferedInputStream.read()).isEqualTo(100)
  }

  @Test
  @Throws(IOException::class)
  fun testSkip2() {
    var i = 0
    while (i < 256) {
      Assertions.assertThat(pooledByteArrayBufferedInputStream.read()).isEqualTo(i)
      i += (pooledByteArrayBufferedInputStream.skip(7) + 1).toInt()
    }
  }

  @Test
  fun testMark() {
    Assertions.assertThat(pooledByteArrayBufferedInputStream.markSupported()).isFalse()
  }

  @Test
  @Throws(IOException::class)
  fun testReadWithByteArray() {
    val readBuffer = ByteArray(5)
    Assertions.assertThat(pooledByteArrayBufferedInputStream.read(readBuffer)).isEqualTo(5)
    assertFilledWithConsecutiveBytes(readBuffer, 0, 5, 0)
  }

  @Test
  @Throws(IOException::class)
  fun testNonFullRead() {
    val readBuffer = ByteArray(200)
    Assertions.assertThat(pooledByteArrayBufferedInputStream.read(readBuffer)).isEqualTo(10)
    assertFilledWithConsecutiveBytes(readBuffer, 0, 10, 0)
    assertFilledWithZeros(readBuffer, 10, 200)
  }

  @Test
  @Throws(IOException::class)
  fun testNonFullReadWithOffset() {
    val readBuffer = ByteArray(200)
    Assertions.assertThat(pooledByteArrayBufferedInputStream.read(readBuffer, 45, 75)).isEqualTo(10)
    assertFilledWithZeros(readBuffer, 0, 45)
    assertFilledWithConsecutiveBytes(readBuffer, 45, 55, 0)
    assertFilledWithZeros(readBuffer, 55, 200)
  }

  @Test
  @Throws(IOException::class)
  fun testReadsCombined() {
    val readBuffer = ByteArray(5)
    var i = 0
    while (i <= 245) {
      Assertions.assertThat(pooledByteArrayBufferedInputStream.read()).isEqualTo(i)

      Assertions.assertThat(pooledByteArrayBufferedInputStream.read(readBuffer)).isEqualTo(5)
      assertFilledWithConsecutiveBytes(readBuffer, 0, readBuffer.size, i + 1)

      Assertions.assertThat(pooledByteArrayBufferedInputStream.read(readBuffer, 1, 3)).isEqualTo(3)
      Assertions.assertThat(readBuffer[0]).isEqualTo((i + 1).toByte())
      assertFilledWithConsecutiveBytes(readBuffer, 1, 4, i + 6)
      Assertions.assertThat(readBuffer[4]).isEqualTo((i + 5).toByte())

      Assertions.assertThat(pooledByteArrayBufferedInputStream.skip(2)).isEqualTo(2)

      i += 11
    }

    Assertions.assertThat(pooledByteArrayBufferedInputStream.available()).isEqualTo(256 - i)
  }

  /**
   * Given byte array, asserts that bytes in [startOffset, endOffset) range are all zeroed.
   *
   * @param byteArray The byte array to check
   * @param startOffset The start offset (inclusive)
   * @param endOffset The end offset (exclusive)
   */
  private fun assertFilledWithZeros(byteArray: ByteArray, startOffset: Int, endOffset: Int) {
    for (i in startOffset until endOffset) {
      Assertions.assertThat(byteArray[i]).isEqualTo(0.toByte())
    }
  }

  /**
   * Given byte array, asserts that each byte in (startOffset, endOffset) range has value equal to
   * value of previous byte plus one (mod 255) and byteArray[startOffset] is equal to firstByte.
   *
   * @param byteArray The byte array to check
   * @param startOffset The start offset (inclusive)
   * @param endOffset The end offset (exclusive)
   * @param firstByte The expected value of the first byte
   */
  private fun assertFilledWithConsecutiveBytes(
      byteArray: ByteArray,
      startOffset: Int,
      endOffset: Int,
      firstByte: Int
  ) {
    var expectedByte = firstByte
    for (i in startOffset until endOffset) {
      Assertions.assertThat(byteArray[i]).isEqualTo(expectedByte.toByte())
      expectedByte++
    }
  }
}
